### PR TITLE
My code review changes

### PR DESCRIPTION
**Pull request by damrondrake**
Pull request for code review changes

### CODE REVIEW

**Comment by damrondrake** [comment]
Description of project:
The card collection app allows users to maintain a database of their own cards that can be searchable. Allowing users to add cards that are not currently in the database. 

Analysis of program:
This program is a Spring Boot application for managing a collection of Card objects, with both front-end pages and back-end REST API endpoints to handle card data. The DissCardController processes requests, and the CardServiceStub acts as a mock service layer that simulates database access. The program is pretty well-organized, keeping controllers, services, and data access separated, but it could benefit from better error handling and more consistent dependency injection.

What I learned:
1.) Constructor Injection vs. Field Injection: 
Using constructor injection instead of field injection is better for testing and keeps dependencies immutable, which is good for reliability.

2.) Error Handling in APIs: 
It’s important to handle errors properly in REST APIs. For example, returning 404 for “not found” or 500 for server errors makes the API responses more meaningful.

3.) Separation of Concerns: 
Separating controller, service, and data access logic keeps the code cleaner and easier to maintain. It makes it easier to make changes in one area without breaking other parts.

Was the program available on time?:
Yes the code was available on time.

Is the program understandable and does it compile:
I would say for the most part that the program is easy to follow and readable and yes, the program does compile.

Rationale behind changes:
1.) Constructor injection for ICardService:
Constructor injection is a best practice in Spring that enhances testability and promotes immutability by ensuring that the ICardService dependency is set at object creation and cannot be changed afterward. This approach simplifies unit testing by allowing for easy mocking of the service, resulting in cleaner and more maintainable code.
2.) Remove explicit HttpHeaders:
Using the produces attribute in the @GetMapping and @PostMapping annotations simplifies the specification of JSON response content types, reducing boilerplate code and enhancing readability. This approach minimizes potential errors related to manual header management, as the framework automatically handles the correct content type.
3.) Added null checks for fetchCardById and fetchCardsByKeyword:
Implementing null checks enhances error handling by returning a 404 NOT FOUND status when a requested card is missing or search results are empty. This approach provides clear feedback to clients, improving the user experience and adhering to RESTful API best practices, making it easier for front-end developers to manage responses.

File: src/main/java/com/teamfive/disscard/DissCardController.java [hunk 1]
```diff
@@ -32,23 +35,28 @@ public class DissCardController {
...
+    private final ICardService cardService;
 
-    Logger log = LoggerFactory.getLogger(this.getClass());
+    private final Logger log = LoggerFactory.getLogger(this.getClass());
+
+    // Constructor injection for better testability and immutability
+    @Autowired
+    public DissCardController(ICardService cardService) {
+        this.cardService = cardService;
+    }
```
Comment (damrondrake) [inline src/main/java/com/teamfive/disscard/DissCardController.java:46]: Constructor injection for better testability.
Source: https://martinfowler.com/articles/microservices.html

File: src/main/java/com/teamfive/disscard/DissCardController.java [hunk 2]
```diff
@@ -58,54 +66,54 @@ public String index(Model model) {
      * Fetches a list of all cards registered with DissCard
      * @return A list of JSON objects representing cards
      */
-    @GetMapping("/card/list")
+    @GetMapping(value = "/card/list", produces = MediaType.APPLICATION_JSON_VALUE)
```
Comment (damrondrake) [inline src/main/java/com/teamfive/disscard/DissCardController.java:69]: Use produces attribute to specify response type as JSON:
Source: https://www.tutorialspoint.com/java/java_class_attributes.htm

File: src/main/java/com/teamfive/disscard/DissCardController.java [hunk 3]
```diff
@@ -58,54 +66,54 @@ public String index(Model model) {
...
+    @GetMapping(value = "/card/search/{keyword}", produces = MediaType.APPLICATION_JSON_VALUE)
     public ResponseEntity<List<Card>> fetchCardsByKeyword(@PathVariable("keyword") String keyword) {
         List<Card> fetchedCards = cardService.searchByName(keyword);
-        HttpHeaders headers = new HttpHeaders();
-        headers.setContentType(MediaType.APPLICATION_JSON);
-        return new ResponseEntity<>(fetchedCards, headers, HttpStatus.OK);
+        if (fetchedCards == null || fetchedCards.isEmpty()) {
+            return new ResponseEntity<>(HttpStatus.NOT_FOUND);
+        }
+        return new ResponseEntity<>(fetchedCards, HttpStatus.OK);
```
Comment (damrondrake) [inline src/main/java/com/teamfive/disscard/DissCardController.java:100]: Added null check to handle case where card is not found and empty check for search results:
Source: https://www.tutorialspoint.com/java/java_class_attributes.htm

File: src/main/java/com/teamfive/disscard/DissCardController.java [hunk 4]
```diff
@@ -58,54 +66,54 @@ public String index(Model model) {
...
     }
 
     /**
      * Adds a card to DissCard's systems
      * @param card Card to be added
-     * @return A JSON object representing the card that was added
+     * @return A JSON object representing the card that was added or an error response if saving fails
      */
-    @PostMapping(value="/card/add", consumes="application/json", produces="application/json")
+    @PostMapping(value = "/card/add", consumes = MediaType.APPLICATION_JSON_VALUE, produces = MediaType.APPLICATION_JSON_VALUE)
```
Comment (damrondrake) [inline src/main/java/com/teamfive/disscard/DissCardController.java:108]: Simplified response handling, directly returning the new card.
Source: https://www.tutorialspoint.com/java/java_class_attributes.htm

**Comment by leichtje** [comment]
Made the corrections to deconflict the merge request and Pulled into the project

**Pull request merged**